### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by phoenix128**

### CODE REVIEW

File: Model/Provider/Engine/Authy.php [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+        return false;
+    }
+
+    /**
+     * Enroll in Authy
+     * @param UserInterface $user
+     * @return bool
+     * @throws LocalizedException
+     */
+    public function enroll(UserInterface $user)
```
Comment (vkublytskyi) [inline Model/Provider/Engine/Authy.php]: Combing in class methods that are required by EngineInterface and engine specific methods lead to very big classes that are hard to read and change. Seems like all public methods of this class are pretty decoupled from each other so separation them to individual classes may improve code readability and simplify maintenance in future.

File: view/adminhtml/web/js/u2f.js [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file or at
+// https://developers.google.com/open-source/licenses/bsd
+
+/* eslint-disable */
+// jscs:disable
+
+/**
+ * @fileoverview The U2F api.
```
[the remark this answers is not in the document]
Comment (phoenix128) [reply view/adminhtml/web/js/u2f.js]: Unfortunately it seems to be not available in any CDN.

File: Model/Provider/Engine/Authy.php (hunk 1 above)
Comment (phoenix128) [reply Model/Provider/Engine/Authy.php]: Refactored in 
https://github.com/magento/magespecialist_TwoFactorAuth/pull/1/commits/5a4aa8971b19ee3b4479c461beff4eba03d62004

File: Observer/ViewBlockAbstractToHtmlBefore.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        if (empty($this->tfa->getAllEnabledProviders())) {
+            return;
+        }
+
+        /** @var $block \Magento\User\Block\User\Edit\Tabs */
+        $block = $observer->getBlock();
+
+        $nameInLayout = $block->getNameInLayout();
+        if ($nameInLayout == 'adminhtml.user.edit.tabs') {
```
[the remark this answers is not in the document]
Comment (phoenix128) [reply Observer/ViewBlockAbstractToHtmlBefore.php]: I agree. But I have to keep it compliant with previous Magento versions.
I reworked it using a plugin to that specific tab class. In this way I will avoid wasting time checking each single block.

Check: https://github.com/magento/magespecialist_TwoFactorAuth/pull/1/commits/76c17e38cc02ed4625d8d4b7a5a43ecc121507ba

File: view/adminhtml/templates/msp_twofactorauth/provider/u2fkey/configure.phtml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+        <div id="u2f-login">
+            <img class="loading" src="<?php echo $block->getViewFileUrl('MSP_TwoFactorAuth::images/logging.gif') ?>"/>
+            <h3><?php echo __('Access granted, redirecting you to Magento...') ?></h3>
+        </div>
+
+    </fieldset>
+</form>
+
+<script>
```
[the remark this answers is not in the document]
Comment (phoenix128) [reply view/adminhtml/templates/msp_twofactorauth/provider/u2fkey/configure.phtml]: I completely refactored authentication dialogs with UI components, non more inline scripts :P .
See: https://github.com/magento/magespecialist_TwoFactorAuth/pull/1/commits/87ea1c222006bf31d74611775df2abd0b1375d39

File: Observer/AdminUserLoadAfter.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    /**
+     * @param Observer $observer
+     * @return void
+     */
+    public function execute(Observer $observer)
+    {
+        $user = $observer->getEvent()->getObject();
+        $user->setData('msp_tfa_providers', $this->userConfigManager->getProvidersCodes($user->getId()));
```
[the remark this answers is not in the document]
Comment (phoenix128) [reply Observer/AdminUserLoadAfter.php]: Yes, but AFAIK, `Magento\User\Api\Data\UserInterface` does not implement `ExtensibleDataInterface`. I think I cannot do that in this case. Am I missing some info?

File: Command/TfaReset.php [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        $this->userResource->load($user, $userName, 'username');
+        if (!$user->getId()) {
+            throw new LocalizedException(__('Unknown user %1', $userName));
+        }
+
+        $provider = $this->providerPool->getProviderByCode($providerCode);
+
+        $this->userConfigManager->resetProviderConfig($user->getId(), $providerCode);
+
+        $output->writeln('' . __('Provider %1 has ben reset for user %2', $provider->getName(), $userName));
```
Comment (TomashKhamlai) [inline Command/TfaReset.php]: 'Provider %1 has been reset for user %2'

Comment (phoenix128) [reply Command/TfaReset.php]: My PHP dyslexia :P

File: etc/module.xml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * @category   MSP
+ * @package    MSP_TwoFactorAuth
+ * @copyright  Copyright (c) 2017 Skeeller srl (http://www.magespecialist.it)
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+-->
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
+    <module name="MSP_TwoFactorAuth" setup_version="2.1.8">
```
Comment (TomashKhamlai) [inline etc/module.xml]: Please remove setup_version attribute just like it was done [here](https://github.com/magento/magento2-sample-data/commit/9897dd69b8239b34d8833c3ad9e9c60a8b1a425a)

Comment (vkublytskyi) [reply etc/module.xml]: @phoenix128 you may ignore the previous comment. This requirement related only to core modules and can not be applied if you would like to support Magento 2.1 or 2.2. Issue with installation will be fixed soon by internal team.

**Pull request merged**